File: TypeScript/roman2int/13.ts
function romanToInt(s: string): number {
  let result = 0;
  let number = 0;
  for (let i = s.length - 1; i >= 0; --i) {
    switch (s[i]) {
      case "I":
        number = 1;
        break;
      case "V":
        number = 5;
        break;
      case "X":
        number = 10;
        break;
      case "L":
        number = 50;
        break;
      case "C":
        number = 100;
        break;
      case "D":
        number = 500;
        break;
      case "M":
        number = 1000;
        break;
    }
    if (number * 4 < result) {
      result -= number;
    } else {
      result += number;
    }
  }
  return result;
}

export default romanToInt;

File: Golang/roman_to_integer/13.go
package romantointeger

func romanToInt(s string) int {
	romanMap := map[string]int{
		"I": 1,
		"V": 5,
		"X": 10,
		"L": 50,
		"C": 100,
		"D": 500,
		"M": 1000,
	}

	result := 0

	// Reverse `s`
	func(x *string) {
		r := ""
		for _, v := range *x {
			r = string(v) + r
		}
		*x = r
	}(&s)

	for _, c := range s {
		number, ok := romanMap[string(c)]
		if !ok {
			return -1
		}

		if number*4 < result {
			result -= number
		} else {
			result += number
		}
	}

	return result
}
